cores/rp2040/main.cpp

```diff
@@ -80,7 +81,36 @@ static struct _reent *_impure_ptr1 = nullptr;
 
 extern "C" int main() {
 #if F_CPU != 125000000
-    set_sys_clock_khz(F_CPU / 1000, true);
+
+    // Optimal core frequency (under and overvolting) based on
+    // https://forums.raspberrypi.com/viewtopic.php?p=1820686&sid=a1bc42fab114da743f958a5dbf160925#p1820686
+    uint32_t frequency_khz = F_CPU / 1000;
+    if (frequency_khz <= 120000) {
```

Comment:
I'm very worried about changing the voltage for in-spec operation.  The RPI guys picked a number after, I assume, some serious QC and binning to ensure their boards were solid over the entire temp range and component variations.  That post doesn't include anything about temperature schmooing, and seems to be only on a single part.

What I've found more limiting is the flash SPI speeds at these crazy high multipliers (we're making the [Celeron 300A](https://www.tomshardware.com/picturestory/636-best-overclocking-cpu.html) look like a snail!).  The QSPI is taken as a divisor of the system clock, so there's a 1:1 increase in it and eventually the flash chip doesn't operate properly.

Would you be able to bracket this voltage tweaking under a menu item in the `generic` device?  Since you'll need to tweak the flash divisor in most cases at these high multipliers, an `Over/Undervolt` menu with a `Default Voltage` and an `Auto Over/Undervolt` item would let users take advantage of this

Reply:
Another possible option would be a `Voltage` menu with the items `Default` and `Auto` (uses the logic here), `0.85V (<=120MHz)`, `0.90V (<=150MHz)`, ... so that the user would be able to manually select one but have some guidance.

Reply:
One more gotcha, when determining the Vin for a device you normally run a test app that turns on all peripherals and burns as much power as possible (since as the I increases, the V on-chip will drop because of the power plane resistance).  His test wasn't doing anything like that AFAICT.  So it would give a lower functional Vin than a fully loaded chip...

Reply:
> Another possible option would be a `Voltage` menu with the items `Default` and `Auto` (uses the logic here), `0.85V (<=120MHz)`, `0.90V (<=150MHz)`, ... so that the user would be able to manually select one but have some guidance.

I would prefer a method where frequencies above 270 MHz automatically set the core voltage and the flash clock divider, otherwise it will not be very user friendly. With my DSP examples for the Datanoise PicoADK, the core is required to run at 402 MHz.

Reply:
1. I still like the idea of a manual `Voltage` menu on the `generic` board with `Default` being the SDK standard and everything else up to the user's prerogative.  (Set it to 0.8 and 300MHZ, watch your board do nothing...)
2. Adding the 2 add'l frequencies is no big deal and I have no issue with.  Leave it at `Overclock` and we're good.
3. For your specific case, though, using that same infrastructure, you could set the Datanoise board's default clock (it's a define you can set in makeboards.py) and the voltage (would be a new define introduced in 1).  Then your board would "just work" OOTB at the proper frequency without affecting anything else.

For selecting dividers, it's not quite so simple because right now they're just `.S` files that `makeboards` has no idea about other than some ASCII strings for their names.  It might be possible to have it parse the names, but really you should be specifying the proper .S file for your 400mhz operation at 1.3V (or whatever you chose).

I suppose it would be theoretically possible to look thru the datasheets of all the supported flash chips and see what the max QSPI frequency is, but in reality that isn't enough since board design comes in to play.  If there's too much capacitance or induced noise affecting setup/hold, the datasheet #s wouldn't be achievable.  So figuring out where to silently adjust is not so easy.  I'd rather punt that to the overclocker....it's always fun playing w/memory timings in the BIOS, right? :laughing: